clippy_lints/src/needless_bool.rs

```diff
@@ -138,76 +138,83 @@ impl LintPass for BoolComparison {
-                        BOOL_COMPARISON,
-                        e.span,
-                        "equality checks against true are unnecessary",
-                        "try simplifying it as shown",
-                        hint.to_string(),
-                        applicability,
-                    );
-                },
-                (Other, Bool(true)) => {
-                    let hint = snippet_with_applicability(cx, left_side.span, "..", &mut applicability);
-                    span_lint_and_sugg(
-                        cx,
-                        BOOL_COMPARISON,
-                        e.span,
-                        "equality checks against true are unnecessary",
-                        "try simplifying it as shown",
-                        hint.to_string(),
-                        applicability,
-                    );
-                },
-                (Bool(false), Other) => {
-                    let hint = Sugg::hir_with_applicability(cx, right_side, "..", &mut applicability);
-                    span_lint_and_sugg(
-                        cx,
-                        BOOL_COMPARISON,
-                        e.span,
-                        "equality checks against false can be replaced by a negation",
-                        "try simplifying it as shown",
-                        (!hint).to_string(),
-                        applicability,
-                    );
-                },
-                (Other, Bool(false)) => {
-                    let hint = Sugg::hir_with_applicability(cx, left_side, "..", &mut applicability);
-                    span_lint_and_sugg(
-                        cx,
-                        BOOL_COMPARISON,
-                        e.span,
-                        "equality checks against false can be replaced by a negation",
-                        "try simplifying it as shown",
-                        (!hint).to_string(),
-                        applicability,
-                    );
-                },
+        if let ExprKind::Binary(Spanned { node, .. }, ..) = e.node {
+            match node {
+                BinOpKind::Eq => check_comparison(
+                    cx,
+                    e,
+                    "equality checks against true are unnecessary",
+                    "equality checks against false can be replaced by a negation",
+                    |h| h,
+                    |h| !h,
+                ),
+                BinOpKind::Ne => check_comparison(
+                    cx,
+                    e,
+                    "inequality checks against true can be replaced by a negation",
+                    "inequality checks against false are unnecessary",
+                    |h| !h,
+                    |h| h,
+                ),
                 _ => (),
             }
         }
     }
 }
 
+fn check_comparison<'a, 'tcx, F, G>(
+    cx: &LateContext<'a, 'tcx>,
+    e: &'tcx Expr,
+    true_message: &str,
+    false_message: &str,
+    true_hint: F,
```

Comment:
just a nit: remove the generic parameters and replace their uses in the argument list with `impl FnOnce(Sugg<'_>) -> Sugg<'_>`